File: solutions/typescript/dnd-character/1/dnd-character.ts
export class DnDCharacter {
  // Thuộc tính ability scores
  strength: number;
  dexterity: number;
  constitution: number;
  intelligence: number;
  wisdom: number;
  charisma: number;
  hitpoints: number;

  constructor() {
    // Tạo 6 ability scores ngẫu nhiên
    this.strength = DnDCharacter.generateAbilityScore();
    this.dexterity = DnDCharacter.generateAbilityScore();
    this.constitution = DnDCharacter.generateAbilityScore();
    this.intelligence = DnDCharacter.generateAbilityScore();
    this.wisdom = DnDCharacter.generateAbilityScore();
    this.charisma = DnDCharacter.generateAbilityScore();

    // Hitpoints = 10 + constitution modifier
    this.hitpoints = 10 + DnDCharacter.getModifierFor(this.constitution);
  }

  // Hàm tung xúc xắc 6 mặt (1–6)
  private static rollDie(): number {
    return Math.floor(Math.random() * 6) + 1;
  }

  // Hàm tạo ability score: roll 4d6, bỏ nhỏ nhất, cộng 3 cái còn lại
  public static generateAbilityScore(): number {
    const rolls = [
      this.rollDie(),
      this.rollDie(),
      this.rollDie(),
      this.rollDie(),
    ];
    rolls.sort((a, b) => a - b); // sắp xếp tăng dần
    rolls.shift(); // bỏ số nhỏ nhất
    return rolls.reduce((sum, val) => sum + val, 0);
  }

  // Hàm tính modifier từ ability value
  public static getModifierFor(abilityValue: number): number {
    if (abilityValue < 3) {
      throw new Error("Ability scores must be at least 3");
    }
    if (abilityValue > 18) {
      throw new Error("Ability scores can be at most 18");
    }
    return Math.floor((abilityValue - 10) / 2);
  }
}

File: solutions/java/dnd-character/1/src/main/java/DnDCharacter.java
import java.util.*;
import java.util.concurrent.ThreadLocalRandom;

class DnDCharacter {
    private final int strength;
    private final int dexterity;
    private final int constitution;
    private final int intelligence;
    private final int wisdom;
    private final int charisma;
    private final int hitpoints;

    public DnDCharacter() {
        this.strength = ability(rollDice());
        this.dexterity = ability(rollDice());
        this.constitution = ability(rollDice());
        this.intelligence = ability(rollDice());
        this.wisdom = ability(rollDice());
        this.charisma = ability(rollDice());
        this.hitpoints = 10 + modifier(this.constitution);
        
    }
    
    int ability(List<Integer> scores) {
        int sum = 0;
        int min = Integer.MAX_VALUE;
        for(int v : scores) {
            sum += v;
            if(v < min) min = v;
        }
        return sum -min;
    }

    List<Integer> rollDice() {
        List<Integer> rolls = new ArrayList<>(4);
        for(int i = 0; i < 4; i++) {
            rolls.add(ThreadLocalRandom.current().nextInt(1, 7));
        }
        return rolls;
    }

    int modifier(int input) {
        return Math.floorDiv(input - 10, 2);
    }

    int getStrength() {
        return strength;
    }

    int getDexterity() {
        return dexterity;
    }

    int getConstitution() {
        return constitution;
    }

    int getIntelligence() {
        return intelligence;
    }

    int getWisdom() {
        return wisdom;
    }

    int getCharisma() {
       return charisma;
    }

    int getHitpoints() {
        return hitpoints;
    }
}
